src/Orleans.Core.Abstractions/Serialization/GrainReferenceKeyInfo.cs

```diff
@@ -0,0 +1,212 @@
+using System;
+using System.Net;
+using Orleans.Runtime;
+
+namespace Orleans.Serialization
+{
+    /// <summary>
+    /// type is a low level representation of graui reference keys to enable
+    /// space-efficient serialization of grain references.
+    /// </summary>
+    /// <remarks>
+    /// This type is not intended for general use. It's a highly specialized type
+    /// for serializing and deserializing grain references, and as such is not
+    /// generally something you should pass around in your application.
+    /// </remarks>
+    public readonly struct GrainReferenceKeyInfo
+    {
+        /// <summary>
+        /// Observer id. Only applicable if <see cref="HasObserverId"/> is true.
+        /// </summary>
+        /// <remarks>In most cases, is Guid.Empty.</remarks>
+        public Guid ObserverId { get; }
+
+        /// <summary>
+        /// Target silo. Only applicable if <see cref="HasTargetSilo"/> is true.
+        /// </summary>
+        /// <remarks>In most cases, is <c>(null, 0)</c>.</remarks>
+        public (IPEndPoint endpoint, int generation) TargetSilo { get; }
+
+        /// <summary>
+        /// Generic argument. Only applicable if <see cref="HasGenericArgument"/> is true.
+        /// </summary>
+        /// <remarks>In most cases, is <c>null</c>.</remarks>
+        public string GenericArgument { get; }
+
+        /// <summary>
+        /// Grain key. For more specialized views, see <c>KeyAs*</c> methods and <c>KeyFrom*</c>.
+        /// </summary>
+        public (ulong, ulong, ulong, string) Key { get; }
+
+        /// <summary>
+        /// Whether or not key info has observer id.
+        /// </summary>
+        public bool HasObserverId => ObserverId != Guid.Empty;
+
+        /// <summary>
+        /// Whether or not key info has silo endpoint.
+        /// </summary>
+        public bool HasTargetSilo => TargetSilo.endpoint != null;
+
+        /// <summary>
+        /// Whether or not key info has generic argument.
+        /// </summary>
+        public bool HasGenericArgument => GenericArgument != null;
+
+        /// <summary>
+        /// Whether or not key info has a long key.
+        /// </summary>
+        public bool IsLongKey => Key.Item1 == 0;
+
+        /// <summary>
+        /// Whether or not key info has key extension.
+        /// </summary>
+        public bool HasKeyExt
+        {
+            get
+            {
+                var category = UniqueKey.GetCategory(Key.Item3);
+                return category == UniqueKey.Category.KeyExtGrain
+                    || category == UniqueKey.Category.GeoClient; // geo clients use the KeyExt string to specify the cluster id
+            }
+        }
+
+        public (Guid key, ulong typeCode) KeyAsGuid()
```

Comment:
Just a note like before that while looking self explanatory, I wonder if should be commented too. Probably not, but for someone else to decide.